database/controller.go

```diff
@@ -0,0 +1,74 @@
+package database
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+)
+
+// ErrNotSupported is returned when an optional method is not supported by the Store implementation.
+var ErrNotSupported = errors.New("not supported")
+
+// A StoreController is used by the goose package to interact with a database. This type is a
+// wrapper around the Store interface, but can be extended to include additional (optional) methods
+// that are not part of the core Store interface.
+type StoreController struct {
+	store Store
+}
+
+var _ Store = (*StoreController)(nil)
+
+// NewStoreController returns a new StoreController that wraps the given Store.
+//
+// If the Store implements the following optional methods, the StoreController will call them as
+// appropriate:
+//
+//   - TableExists(context.Context, DBTxConn) (bool, error)
+//
+// If the Store does not implement a method, it will either return a [ErrNotSupported] error or fall
+// back to the default behavior.
+func NewStoreController(store Store) *StoreController {
```

Comment:
The downside with this pattern is external implementations don't have a type assertion to lean on, and so it's really easy to _believe_ a method is implemented, but in reality the method signature doesn't match. You lose that nice compile-time check.

Another pattern could be to have a second interface like `StoreExtender` that's a superset of `Store`.

```go
type StoreExtender interface {
	Store
	// TableExists is an optional method that checks if the version table exists in the database. It is
	// recommended to implement this method if the database supports it, as it can be used to optimize
	// certain operations.
	TableExists(ctx context.Context, db DBTxConn) (bool, error)
}
```

There's pros/cons to both approaches, need to sleep on this a bit and chat with folks to get feedback.